include/boost/winapi/critical_section.hpp

```diff
@@ -20,89 +20,97 @@
 
 extern "C" {
 #if !defined( BOOST_WINAPI_IS_MINGW )
+
+// Windows CE uses a different name for the structure
+#if defined (_WIN32_WCE)
+struct CRITICAL_SECTION;
+typedef CRITICAL_SECTION BOOST_WINAPI_RTL_CRITICAL_SECTION;
```

Comment:
This suggestion, if implemented literally, causes 3 issues.
1.  public boost::winapi interface mentions detail namespace
2.  it create illegal overload with extern "C" linkage
3. it causes ambiguity if unqualified call (like error C2668: 'EnterCriticalSection': ambiguous call to overloaded function)

I can amend the commit to exclude the change about typedef for _RTL_CRITICAL_SECTION altogether, and open separate PR only about that. What do you think?

Reply:
1. Not sure I understand what you mean. Public declarations in `boost::winapi` do not and should not use neither `::_RTL_CRITICAL_SECTION` nor `::CRITICAL_SECTION` but a dedicated structure `boost::winapi::_RTL_CRITICAL_SECTION`. The typedef I suggested (let's call it `boost::winapi::detail::winsdk_critical_section`) will only be used in the declarations in the global namespace - i.e. those mirroring the ones from Windows SDK.
2. Since the typedef will refer to the same structure that is defined in the Windows SDK, this will not be an overload but rather a declaration of the same function.
3. Unqualified calls from within Boost should not happen since all Boost.WinAPI public symbols are in `boost::winapi`. Any Boost library that is using Boost.WinAPI has to specify at the very least `winapi` namespace. Unqualified calls from non-Boost user's code will resolve to the function from the Windows SDK (see point 2; the declaration in Boost.WinAPI must be equivalent to that in Windows SDK and thus not introduce an overload).

Reply:
ok, it works, sorry.

3 more questions.

1.
Do you prefer boost::winapi::detail::winapi_critical_section
or boost::winapi::detail::winsdk_critical_section ?

2. 
I can apply same approach for MINGW #ifdef in the same file for consistency.

3.
The internal layout of the CRITICAL_SECTION  on CE is different from NT. Fortunately, sizeof of CE variant is smaller.

boost::winapi::_RTL_CRITICAL_SECTION mimics NT. Do we want to do something about it?

This is CE variant:

`typedef` struct CRITICAL_SECTION {
    unsigned int LockCount;         /* Nesting count on critical section */
    HANDLE OwnerThread;         	/* Handle of owner thread */
    HANDLE hCrit;					/* Handle to this critical section */
    DWORD needtrap;					/* Trap in when freeing critical section */
    DWORD dwContentions;			/* Count of contentions */
} CRITICAL_SECTION, `*LPCRITICAL_SECTION;`

Reply:
1. `winsdk_critical_section` is better as it implies that it refers to the type defined in Windows SDK rather than Boost.WinAPI.
2. Yes, that would be great, thanks. Do check that other than the structure name the function signatures are the same, though.
3. I think it would be better to mimic the CE structure on Windows CE.